### PR TITLE
Two theorems about non-semiregular almost discrete spaces, plus a fixed trait

### PR DESCRIPTION
**Pull request by danflapjax**
Interestingly, being almost discrete and not semiregular fully characterizes a space up to cardinality, so we can have some very strong theorems involving the two properties.

### CODE REVIEW

**Comment by GeoffreySangston** [comment]
I think it's all correct. Do you have any suggestions for checking if there are now any removable theorems (not suggesting there are, but just for future reference)?

**Comment by GeoffreySangston** [comment]
I think some external reference would be better, because the theorems do have at least twice as much text as the average theorem in pi-base. But if others disagree please override.

**Comment by prabau** [comment]
I agree this is too long.  Can this be some mathse posts?

**Comment by prabau** [comment]
@danflapjax Is this one of the theorems that were pending from #821?

**Comment by prabau** [comment]
@danflapjax Reminder: as a result of #821 which did not get implemented, we now have a gap of available ids for theorems: T603 - T607.  If you are planning to add more results about almost discrete, maybe you can use some of these ids.

**Comment by danflapjax** [comment]
I'm not really sure how to even make it a worthwhile MathSE question, but I can try. Showing that an almost discrete space that isn't semiregular is the disjoint union of the Sierpinski space and a discrete space I guess? It feels quite basic to me even if it is a bit wordy.

> ... the theorems do have at least twice as much text as the average theorem in pi-base.

I was curious if this was literally true; I guess it is: https://observablehq.com/d/489b89d8b5eca5d5

> Is this one of the theorems that were pending from #821?

One of the two is.

**Comment by prabau** [comment]
> Showing that an almost discrete space that isn't semiregular is the disjoint union of the Sierpinski space and a discrete space

Or you can ask an open ended question: What do almost discrete spaces that are not semiregular look like, and wait for answers.

**Comment by yhx-12243** [comment]
> I'm not really sure how to even make it a worthwhile MathSE question, but I can try. Showing that an almost discrete space that isn't semiregular is the disjoint union of the Sierpinski space and a discrete space I guess? It feels quite basic to me even if it is a bit wordy.
> 
> > ... the theorems do have at least twice as much text as the average theorem in pi-base.
> 
> I was curious if this was literally true; I guess it is: https://observablehq.com/d/489b89d8b5eca5d5
> 
> > Is this one of the theorems that were pending from #821?
> 
> One of the two is.

I've tested. This PR is completely independent of #821.

**Comment by Moniker1998** [comment]
Any updates on this?

**Comment by danflapjax** [comment]
> Any updates on this?

The main update is that I graduated from university, and I've been busy with other things like spending holiday time with my family. :)

This PR should be ready for review now.

**Comment by prabau** [comment]
@danflapjax Congratulations!

**Comment by prabau** [comment]
@danflapjax Reading your mathse post.  Nice proof.

Just very minor details.  In the second paragraph, "Any neighborhood of the non-isolated point p is closed", you mean "open neighborhood", right?

Also, you never said it explicitly I think, but shouldn't it be mentioned at least once that the Sierpinski space union discrete space is almost discrete and not semiregular?  (no proof needed, it's easy to check).

And in the definition of almost discrete in the question, I think you meant to say "there exists a non-isolated point $p\in X$ such that ...".

**Comment by prabau** [comment]
That fixed the extremally disconnected trait for S188. Good deal.

**Comment by prabau** [comment]
S188: you removed P36 (connected), but the trait can't be derived.

**Comment by yhx-12243** [comment]
Alright, in my local version it requires traits on #821.

But that project was holding for a long time.

external traits

**Comment by prabau** [comment]
@yhx-12243 Ah, yes.  So you think it would resolve itself if we just merge this to `main`?

**Comment by yhx-12243** [comment]
> @yhx-12243 Ah, yes. So you think it would resolve itself if we just merge this to `main`?

In my opinion, I think traits in #821 will finally resolved at some time in future, and this trait can be deduced by them.

However, it is a possible choice keep {S188|P36} now because we don't know when traits in #821 will be merged.

**Comment by prabau** [comment]
Yes, I think we should keep that trait for now.  And then remove it when portions of #821 get added.

**Comment by prabau** [comment]
@yhx-12243 Can you explain?  What exactly do you mean by "steal" that trait?

I have not looked at #821 in detail for a long time.  What was that PR doing with P188?

**Comment by yhx-12243** [comment]
#821 has a trait namely "Almost + ~Semiregular + Cardinality $\geq 3$ => ~Connected", as you can see, this trait is closely related to this PR.

So I think instead of putting {S188|P36} back we can just add this trait to this PR and leaving other trait in #821.

**Comment by prabau** [comment]
@yhx-12243 I see.  What confused me was the word "trait".  A **trait** of a space is just a property of a space being true or false.  Some traits are derived, and some traits are asserted in a **trait file**.  This is completely different from a **theorem**, which seems is what you meant.

To keep things manageable and moving along, I would recommend to rather restore the P36 trait for S188, and approve this PR.  That will allow to merge this old PR.

And then you can write a separate PR with that new theorem, which we can review and approve quickly.  And the P36 trait for S188 can be removed as redundant at the same time.

Would that be ok with you?

**Comment by yhx-12243** [comment]
> @yhx-12243 I see. What confused me was the word "trait". A **trait** of a space is just a property of a space being true or false. Some traits are derived, and some traits are asserted in a **trait file**. This is completely different from a **theorem**, which seems is what you meant.

Sorry it's my fault.

> To keep things manageable and moving along, I would recommend to rather restore the P36 trait for S188, and approve this PR. That will allow to merge this old PR.
> 
> And then you can write a separate PR with just that new theorem, which we can review and approve quickly. And the P36 trait for S188 can be removed as redundant at the same time.
> 
> Would that be ok with you?

I think it is okay.

**Comment by prabau** [comment]
@yhx-12243 Great.  I'll restore P36 from the branch now, and then approve and merge.

I wanted to ask, when you talk about your "local version" where you detected the problem, is that a local branch that you have in a local repository or something else you are using?  It seems you are able to detect these kinds of problems very quickly.

**Comment by yhx-12243** [comment]
> @yhx-12243 Great. I'll restore P36 from the branch now, and then approve and merge.
> 
> I wanted to ask, when you talk about your "local version" where you detected the problem, is that a local branch that you have in a local repository or something else you are using? It seems you are able to detect these kinds of problems very quickly.

Just a local copy of this repository and web, and can do locally merge by Git techniques and build the data source on localhost, then change the "Host" field on https://topology.pi-base.org/dev (I think "Host" is just used for this).

**Comment by prabau** [comment]
Very cool.  At this point, I wouldn't be able to run the web system on my own machine.  Too complicated.  But I am glad it works well for you.

**Comment by danflapjax** [comment]
The connectedness trait should be redundant: https://topology.pi-base.org/spaces?q=Almost+discrete+%2B+%7ESemiregular+%2B+Extremally+disconnected+%2B+Cardinality+%24%5Cgeq+3%24+%2B+Connected

**Comment by prabau** [comment]
@yhx-12243 I have restored the P36 (connected) trait and checked that things are correct on the web page.
It actually shows that P36 was redundant, but that's maybe due to some circular dependency?

In any case, approving and merging this right now.  And then feel free to create a separate PR for that new theorem plus cleanup as discussed.  Please just that theorem.  Other theorems can go in separate chunks as needed.

**Comment by yhx-12243** [comment]
> The connectedness trait will be redundant thanks to more recent theorems: https://topology.pi-base.org/spaces?q=Almost+discrete+%2B+%7ESemiregular+%2B+Extremally+disconnected+%2B+Cardinality+%24%5Cgeq+3%24+%2B+Connected

This is right. The original theorem is covered by the stronger one and it wouldn't be added, as well as {S188|P36}, which is not need neither.

**Comment by yhx-12243** [comment]
> The connectedness trait will be redundant thanks to more recent theorems: https://topology.pi-base.org/spaces?q=Almost+discrete+%2B+%7ESemiregular+%2B+Extremally+disconnected+%2B+Cardinality+%24%5Cgeq+3%24+%2B+Connected

Oh no! This is a example of deficiency of deduction system, just like pi-base/web#199.

**Comment by prabau** [comment]
@danflapjax caught me just on time.  I'll remove P36 again, and then approve and merge.

**Comment by yhx-12243** [comment]
We've encountered a situation that with "Connected" we can get contradiction, but without "Connected" we can't deduce it. This is indeed a deficiency of our deduction system, just as said in pi-base/web#199.

**Review by prabau — APPROVED** [review]

**Comment by prabau** [comment]
Well, I merge the PR.  And now https://topology.pi-base.org/spaces/S000188 cannot derive the connected trait.

@danflapjax can you take care of adding it back with another PR, when you get the chance?

**Comment by prabau** [comment]
@StevenClontz you may be interested in the last few commernts above.

**Comment by yhx-12243** [comment]
The basically rationale is that, suppose we have $A \implies B$, $A \implies C$ and $B + C \implies D$.
Then of course $A \wedge \neg D$ is a contradiction. But based on our deduction system, $\neg D$ can't deduce $\neg A$. This is the intrinsic deficiency of boolean constraint propagation algorithm.

Replacing it by SAT solver may completely solve it, by it'll introduce another tradeoff since SAT is a famous NP-complete problem. This is a very awkward situation (would force us to add "seems" redundant theorems).
